src/localsocketcontroller.cpp

```diff
@@ -343,9 +357,104 @@ void LocalSocketController::parseCommand(const QByteArray& command) {
+  // is used to detect a crash or failure of the daemon.
+  if (!expectedResponseType.isEmpty()) {
+    QTimer* t = new QTimer(this);
+    connect(t, &QTimer::timeout, this,
+            [&] { this->errorOccurred(QLocalSocket::SocketTimeoutError); });
+
+    t->setProperty("responseType", QVariant(expectedResponseType));
+    t->setSingleShot(true);
+    t->start(timeout);
+
+    m_responseTimeouts.append(t);
+  }
+
   Q_ASSERT(m_socket);
-  m_socket->write(QJsonDocument(json).toJson(QJsonDocument::Compact));
-  m_socket->write("\n");
+  m_socket->write(payload);
   m_socket->flush();
 }
+
+void LocalSocketController::clearTimeout(const QString& responseType) {
+  for (QTimer* t : m_responseTimeouts) {
+    QVariant qv = t->property("responseType");
+    if (qv.type() != QVariant::String) {
+      continue;
+    }
+    if (qv.toString() != responseType) {
+      continue;
+    }
+    m_responseTimeouts.removeOne(t);
+    delete t;
+  }
+}
+
+void LocalSocketController::clearAllTimeouts() {
+  while (!m_responseTimeouts.isEmpty()) {
+    QTimer* t = m_responseTimeouts.takeFirst();
+    delete t;
+  }
+}
+
+void LocalSocketController::testDaemonCrash() {
+#ifdef MZ_MACOS
+  pid_t pid;
+  socklen_t len = sizeof(pid);
+  int sd = m_socket->socketDescriptor();
+  if (getsockopt(sd, SOL_LOCAL, LOCAL_PEERPID, &pid, &len) < 0) {
+    return;
+  }
+  if ((pid <= 0) || (pid == getpid())) {
+    return;
+  }
+
+  // Create an authorization session.
+  AuthorizationRef authRef;
+  AuthorizationFlags authFlags =
+      kAuthorizationFlagDefaults | kAuthorizationFlagInteractionAllowed |
+      kAuthorizationFlagPreAuthorize | kAuthorizationFlagExtendRights;
+  OSStatus status = AuthorizationCreate(nullptr, kAuthorizationEmptyEnvironment,
+                                        authFlags, &authRef);
+  if (status != errAuthorizationSuccess) {
+    logger.error() << "Failed to acquire authorization:" << status;
+    return;
+  }
+  auto authGuard = qScopeGuard(
+      [&] { AuthorizationFree(authRef, kAuthorizationFlagDefaults); });
+
+  // Acquire execution permissions.
+  AuthorizationItem authItems = {kAuthorizationRightExecute, 0, nullptr, 0};
+  AuthorizationRights authRights = {1, &authItems};
+  status = AuthorizationCopyRights(authRef, &authRights, nullptr, authFlags,
+                                   nullptr);
+  if (status != errAuthorizationSuccess) {
+    logger.error() << "Failed to copy authorization rights:" << status;
+    return;
+  }
+
+  // Execute 'kill' to terminate the daemon as though it crashed.
+  logger.warning() << "Sending SIGSEGV to:" << pid;
+  QByteArray pidString = QString::number(pid).toUtf8();
+  char killpath[] = "/bin/kill";
+  char killsignal[] = "-SEGV";
+  char* const killargs[] = {killsignal, pidString.data(), nullptr};
+
+#  pragma clang diagnostic push
+#  pragma clang diagnostic ignored "-Wdeprecated-declarations"
+  status = AuthorizationExecuteWithPrivileges(
```

Comment:
Looks like [AuthorizationExecuteWithPrivileges](https://developer.apple.com/documentation/security/1540038-authorizationexecutewithprivileg) is deprecated. Are there security concerns with using this?

Reply:
I think Apple's intention for deprecating this API is that they want to encourage developers to split privileged functionality out of their UI processes and move it into daemons instead, which is exactly how we have structured the VPN on MacOS. However, an IPC method to tell the daemon to crash itself seems even worse from a security perspective.

I think part of the concern with using this API is that it just trains users to accept a security login prompt without thinking, and doesn't do anything to describe (or restrict) what applications can do once they are authorized.

However, for the purposes of a testing command which can only be triggered from the inspector, I think this is fine.

Reply:
There is also the danger that at some point in the future Apple may remove this function entirely, in which case we may need to revisit how to test this feature. But lets cross that bridge when we get there.